granulate_utils/logging/__init__.py

```diff
@@ -0,0 +1,230 @@
+        self.buffer: List[str] = []
+        self.lengths: List[int] = []
+        self.head_serial_no = 0
+        self.next_serial_no = 0
+
+    @property
+    def count(self) -> int:
+        return len(self.buffer)
+
+    @property
+    def utilization(self) -> float:
+        return self.total_length / self.max_total_length
+
+    def append(self, item):
+        self.buffer.append(item)
+        self.lengths.append(len(item))
+        self.total_length += len(item)
+        self.next_serial_no += 1
+        self.handle_overflow()
+
+    def handle_overflow(self) -> None:
+        if self.total_length >= self.max_total_length:
+            self.drop(max(1, int(self.overflow_drop_factor * self.count)))
+
+    def drop(self, n: int):
+        assert n > 0, "n must be positive!"
+        if self.count == 0:
+            return
+        if n > self.count:
+            n = self.count
+        self.head_serial_no += n
+        self.total_length -= sum(self.lengths[:n])
+        del self.lengths[:n]
+        del self.buffer[:n]
+
+
+class BatchRequestsHandler(Handler):
+    """
+    logging.Handler that accumulates logs in a buffer and flushes them periodically or when
+    threshold is reached. The logs are transformed into dicts. Requests to the server are serialized
+    in a dedicated thread to reduce congestion on the server.
+    """
+
+    scheme = "https"
+
+    # If Tuple[float, float], then the first value is connection-timeout and the second read-timeout.
+    # See https://docs.python-requests.org/en/latest/api/#requests.request
+    request_timeout: Union[float, Tuple[float, float]] = 1.5
+    stop_timeout: float = 60.0
+
+    def __init__(
+        self,
+        server_address: str,
+        max_message_size=1 * 1024 * 1024,  # 1mb
+        max_total_length=5 * 1024 * 1024,  # 5mb
+        flush_interval=10.0,
+        flush_threshold=0.8,
+        overflow_drop_factor=0.25,
+    ):
+        super().__init__(logging.DEBUG)
+        self.max_message_size = max_message_size  # maximum message size
+        self.capacity = max_total_length  # maximum size of buffer in bytes
+        self.flush_interval = flush_interval  # maximum amount of seconds between flushes
+        self.flush_threshold = flush_threshold  # force flush if buffer size reaches this percentage of capacity
+        self.overflow_drop_factor = overflow_drop_factor  # drop this percentage of messages upon overflow
+        self.messages_buffer = MessagesBuffer(self.capacity, self.overflow_drop_factor)
+        self.stop_event = threading.Event()
+        self.time_fn = time.time
+        self.server_address = server_address
+        self.session = requests.Session()
+        self.last_flush_time = 0.0
+        self.flush_thread = threading.Thread(target=self._flush_loop, daemon=True, name="Logs flusher")
+        self.start()
+
+    def start(self) -> None:
+        """Called to start the flush thread."""
+        self.flush_thread.start()
```

Comment:
Should we recreate the thread in `start`? IMO It's not expected to be restarted. If restart is desired, recreate the entire handler. Otherwise, we'd be invalidating some invariants and dropping messages for no reason. Comment this.

Reply:
No need, but worth to add assert to make sure it is not recreated (and add as test)

Reply:
Added.